File: src/my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
from functools import partial

class AddTwoIntsClientNode(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")
        self.call_add_two_ints_server(6, 7)
        # Using this version, it is easy to call server multiple times
        self.call_add_two_ints_server(1, 5)
        self.call_add_two_ints_server(11, 12)

    def call_add_two_ints_server(self, a, b):
        client = self.create_client(AddTwoInts, 'add_two_ints')

        while not client.wait_for_service(1.0):
            self.get_logger().warn('Waiting for Server Add Two Ints')

        request = AddTwoInts.Request()
        request.a = a
        request.b = b

        future = client.call_async(request)
        future.add_done_callback(partial(self.callback_call_add_two_ints, a=a, b=b))

    def callback_call_add_two_ints(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info(f'{a} + {b} = {response.sum}')
        except Exception as e:
            self.get_logger().error(f'Service call failed {e}')


def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntsClientNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/my_cpp_pkg/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

class AddTwoIntsClientNode : public rclcpp::Node
{
public:
    AddTwoIntsClientNode() : Node("add_two_ints_client")
    {
        // This will not work, because the future.get() function
        // will block the thread from spinning the node, which is
        // required in order to get the response.
        // callAddTwoIntsService(1, 4);

        // The answer is to use another thread
        threads_.push_back(std::thread(std::bind(
                &AddTwoIntsClientNode::callAddTwoIntsService, this, 1, 4)));

        // Using this method you can run the client multiple times
        threads_.push_back(std::thread(std::bind(
                &AddTwoIntsClientNode::callAddTwoIntsService, this, 6, 42)));
    }

    void callAddTwoIntsService(int a, int b)
    {
        // Create a client with the correct datatype and service name
        auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
        // Wait to connect to service
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for the server to be up...");
        }

        // Once service connection is esablished, create a request object.
        auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        // Fill request with desired parameter data
        request->a = a;
        request->b = b;

        // Send the request asyncronously using the client
        auto future = client->async_send_request(request);

        // the get() function will wait and block until we recieve a response
        try
        {
            auto response = future.get();
            RCLCPP_INFO(this->get_logger(), "%d + %d = %d", a, b, (int)response->sum);
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Service call failed");
        }
    }

private:
    std::vector<std::thread> threads_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsClientNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
